File: sdk-libs/token-client/src/instructions/mint_action.rs
use borsh::BorshDeserialize;
use light_client::{
    indexer::Indexer,
    rpc::{Rpc, RpcError},
};
use light_compressed_account::instruction_data::traits::LightInstructionData;
use light_compressible::config::CompressibleConfig;
use light_token_interface::{
    instructions::{
        extensions::{token_metadata::TokenMetadataInstructionData, ExtensionInstructionData},
        mint_action::{
            CompressAndCloseCMintAction, CompressedMintWithContext, DecompressMintAction,
            MintActionCompressedInstructionData, MintToTokenAction, MintToCompressedAction,
            Recipient, RemoveMetadataKeyAction, UpdateAuthority, UpdateMetadataAuthorityAction,
            UpdateMetadataFieldAction,
        },
    },
    state::CompressedMint,
    LIGHT_TOKEN_PROGRAM_ID,
};
use light_ctoken_sdk::compressed_token::{
    create_compressed_mint::{derive_cmint_compressed_address, find_cmint_address},
    mint_action::MintActionMetaConfig,
};
use solana_instruction::Instruction;
use solana_keypair::Keypair;
use solana_pubkey::Pubkey;
use solana_signer::Signer;

// Backwards compatibility types for token-client
#[derive(Debug, Clone, PartialEq)]
pub struct MintToRecipient {
    pub recipient: Pubkey,
    pub amount: u64,
}

/// High-level action types for the mint action instruction (backwards compatibility)
#[derive(Debug, Clone, PartialEq)]
pub enum MintActionType {
    MintTo {
        recipients: Vec<MintToRecipient>,
        token_account_version: u8,
    },
    UpdateMintAuthority {
        new_authority: Option<Pubkey>,
    },
    UpdateFreezeAuthority {
        new_authority: Option<Pubkey>,
    },
    MintToCToken {
        account: Pubkey,
        amount: u64,
    },
    UpdateMetadataField {
        extension_index: u8,
        field_type: u8,
        key: Vec<u8>,
        value: Vec<u8>,
    },
    UpdateMetadataAuthority {
        extension_index: u8,
        new_authority: Pubkey,
    },
    RemoveMetadataKey {
        extension_index: u8,
        key: Vec<u8>,
        idempotent: u8,
    },
    /// Decompress the compressed mint to a CMint Solana account.
    /// CMint is always compressible - rent_payment must be >= 2.
    DecompressMint {
        cmint_bump: u8,
        /// Rent payment in epochs (prepaid). Must be >= 2.
        rent_payment: u8,
        /// Lamports allocated for future write operations (top-up per write).
        write_top_up: u32,
    },
    /// Compress and close a CMint Solana account. The compressed mint state is preserved.
    /// Permissionless - anyone can call if is_compressible() returns true (rent expired).
    CompressAndCloseCMint {
        /// If true, succeed silently when CMint doesn't exist
        idempotent: bool,
    },
}

/// Parameters for creating a new mint
#[derive(Debug, Clone)]
pub struct NewMint {
    pub decimals: u8,
    pub supply: u64,
    pub mint_authority: Pubkey,
    pub freeze_authority: Option<Pubkey>,
    pub metadata: Option<TokenMetadataInstructionData>,
    pub version: u8,
}

/// Parameters for mint action instruction
#[derive(Debug, Clone)]
pub struct MintActionParams {
    pub compressed_mint_address: [u8; 32],
    pub mint_seed: Pubkey,
    pub authority: Pubkey,
    pub payer: Pubkey,
    pub actions: Vec<MintActionType>,
    /// Required if any action is creating a mint
    pub new_mint: Option<NewMint>,
}

/// Creates a mint action instruction that can perform multiple mint operations
pub async fn create_mint_action_instruction<R: Rpc + Indexer>(
    rpc: &mut R,
    params: MintActionParams,
) -> Result<Instruction, RpcError> {
    // Check if we're creating a new mint
    let is_creating_mint = params.new_mint.is_some();

    // Check if DecompressMint action is present
    let has_decompress_mint = params
        .actions
        .iter()
        .any(|a| matches!(a, MintActionType::DecompressMint { .. }));

    // Check if CompressAndCloseCMint action is present
    let has_compress_and_close_cmint = params
        .actions
        .iter()
        .any(|a| matches!(a, MintActionType::CompressAndCloseCMint { .. }));

    // Get address tree and output queue info
    let address_tree_pubkey = rpc.get_address_tree_v2().tree;

    let (compressed_mint_inputs, proof, state_tree_info) = if is_creating_mint {
        let state_tree_info = rpc.get_random_state_tree_info()?;
        // For creating mint: get address proof and create placeholder compressed mint inputs
        let rpc_proof_result = rpc
            .get_validity_proof(
                vec![],
                vec![light_client::indexer::AddressWithTree {
                    address: params.compressed_mint_address,
                    tree: address_tree_pubkey,
                }],
                None,
            )
            .await?
            .value;

        // Create compressed mint data for creation with actual values
        let new_mint = params.new_mint.as_ref().ok_or_else(|| {
            RpcError::CustomError("NewMint parameters required for mint creation".to_string())
        })?;

        let mint_data =
            light_token_interface::instructions::mint_action::CompressedMintInstructionData {
                supply: new_mint.supply,
                decimals: new_mint.decimals,
                metadata: light_token_interface::state::CompressedMintMetadata {
                    version: new_mint.version,
                    mint: find_cmint_address(&params.mint_seed).0.to_bytes().into(),
                    // false for new mint - on-chain sets to true after DecompressMint
                    cmint_decompressed: false,
                    compressed_address: params.compressed_mint_address,
                },
                mint_authority: Some(new_mint.mint_authority.to_bytes().into()),
                freeze_authority: new_mint.freeze_authority.map(|auth| auth.to_bytes().into()),
                extensions: new_mint
                    .metadata
                    .as_ref()
                    .map(|meta| vec![ExtensionInstructionData::TokenMetadata(meta.clone())]),
            };

        let compressed_mint_inputs = CompressedMintWithContext {
            prove_by_index: false, // Use full proof for creation
            leaf_index: 0,         // Not applicable for creation
            root_index: rpc_proof_result.addresses[0].root_index,
            address: params.compressed_mint_address,
            mint: Some(mint_data),
        };

        (
            compressed_mint_inputs,
            rpc_proof_result.proof.0,
            state_tree_info,
        )
    } else {
        // For existing mint: get validity proof for the compressed mint
        let compressed_mint_account = rpc
            .get_compressed_account(params.compressed_mint_address, None)
            .await?
            .value
            .ok_or(RpcError::AccountDoesNotExist(format!(
                "{:?}",
                params.compressed_mint_address
            )))?;

        // Try to deserialize the compressed mint - may be None if CMint is source of truth
        let compressed_mint: Option<CompressedMint> = compressed_mint_account
            .data
            .as_ref()
            .and_then(|d| BorshDeserialize::deserialize(&mut d.data.as_slice()).ok());

        let rpc_proof_result = rpc
            .get_validity_proof(vec![compressed_mint_account.hash], vec![], None)
            .await?
            .value;

        let compressed_mint_inputs = CompressedMintWithContext {
            prove_by_index: rpc_proof_result.accounts[0].root_index.proof_by_index(),
            leaf_index: compressed_mint_account.leaf_index,
            root_index: rpc_proof_result.accounts[0]
                .root_index
                .root_index()
                .unwrap_or_default(),
            address: params.compressed_mint_address,
            mint: compressed_mint.map(|m| m.try_into().unwrap()),
        };

        (
            compressed_mint_inputs,
            rpc_proof_result.proof.into(),
            rpc_proof_result.accounts[0].tree_info,
        )
    };

    // Build instruction data using builder pattern
    let mut instruction_data = if is_creating_mint {
        MintActionCompressedInstructionData::new_mint(
            compressed_mint_inputs.root_index,
            proof.ok_or_else(|| {
                RpcError::CustomError("Proof is required for mint creation".to_string())
            })?,
            compressed_mint_inputs.mint.unwrap().clone(),
        )
    } else {
        MintActionCompressedInstructionData::new(compressed_mint_inputs.clone(), proof)
    };

    // Convert and add actions using builder pattern
    // Collect decompressed token accounts for MintToCToken actions
    let mut ctoken_accounts = Vec::new();
    let mut ctoken_account_index = 0u8;
    let mut has_mint_to_compressed = false;

    for action in params.actions {
        instruction_data = match action {
            MintActionType::MintTo {
                recipients,
                token_account_version,
            } => {
                has_mint_to_compressed = true;
                // Convert MintToRecipient (solana_sdk::Pubkey) to Recipient ([u8; 32])
                let ctoken_recipients: Vec<Recipient> = recipients
                    .into_iter()
                    .map(|r| Recipient::new(r.recipient, r.amount))
                    .collect();
                instruction_data.with_mint_to_compressed(MintToCompressedAction {
                    token_account_version,
                    recipients: ctoken_recipients,
                })
            }
            MintActionType::MintToCToken { account, amount } => {
                // Add account to the list and use its index
                ctoken_accounts.push(account);
                let current_index = ctoken_account_index;
                ctoken_account_index += 1;

                instruction_data.with_mint_to_token(MintToTokenAction {
                    account_index: current_index,
                    amount,
                })
            }
            MintActionType::UpdateMintAuthority { new_authority } => instruction_data
                .with_update_mint_authority(UpdateAuthority {
                    new_authority: new_authority.map(|a| a.to_bytes().into()),
                }),
            MintActionType::UpdateFreezeAuthority { new_authority } => instruction_data
                .with_update_freeze_authority(UpdateAuthority {
                    new_authority: new_authority.map(|a| a.to_bytes().into()),
                }),
            MintActionType::UpdateMetadataField {
                extension_index,
                field_type,
                key,
                value,
            } => instruction_data.with_update_metadata_field(UpdateMetadataFieldAction {
                extension_index,
                field_type,
                key,
                value,
            }),
            MintActionType::UpdateMetadataAuthority {
                extension_index,
                new_authority,
            } => instruction_data.with_update_metadata_authority(UpdateMetadataAuthorityAction {
                extension_index,
                new_authority: new_authority.to_bytes().into(),
            }),
            MintActionType::RemoveMetadataKey {
                extension_index,
                key,
                idempotent,
            } => instruction_data.with_remove_metadata_key(RemoveMetadataKeyAction {
                extension_index,
                key,
                idempotent,
            }),
            MintActionType::DecompressMint {
                cmint_bump,
                rent_payment,
                write_top_up,
            } => instruction_data.with_decompress_mint(DecompressMintAction {
                cmint_bump,
                rent_payment,
                write_top_up,
            }),
            MintActionType::CompressAndCloseCMint { idempotent } => instruction_data
                .with_compress_and_close_cmint(CompressAndCloseCMintAction {
                    idempotent: if idempotent { 1 } else { 0 },
                }),
        };
    }

    // Build account metas configuration
    let mut config = if is_creating_mint {
        MintActionMetaConfig::new_create_mint(
            params.payer,
            params.authority,
            params.mint_seed,
            address_tree_pubkey,
            state_tree_info.queue,
        )
    } else {
        MintActionMetaConfig::new(
            params.payer,
            params.authority,
            state_tree_info.tree,
            state_tree_info.queue,
            state_tree_info.queue,
        )
    };

    // Add tokens_out_queue if there are MintToCompressed actions
    if has_mint_to_compressed {
        config = config.with_mint_compressed_tokens();
    }

    // Add ctoken accounts if any MintToCToken actions were present
    if !ctoken_accounts.is_empty() {
        config = config.with_ctoken_accounts(ctoken_accounts);
    }

    // Add compressible CMint accounts if DecompressMint or CompressAndCloseCMint action is present
    if has_decompress_mint || has_compress_and_close_cmint {
        let (cmint_pda, _) = find_cmint_address(&params.mint_seed);
        // Get config and rent_sponsor from v1 config PDA
        let config_address = CompressibleConfig::ctoken_v1_config_pda();
        let compressible_config: CompressibleConfig = rpc
            .get_anchor_account(&config_address)
            .await?
            .ok_or_else(|| {
                RpcError::CustomError(format!(
                    "CompressibleConfig not found at {}",
                    config_address
                ))
            })?;
        config = config.with_compressible_cmint(
            cmint_pda,
            config_address,
            compressible_config.rent_sponsor,
        );
        // DecompressMint needs mint_signer even when not creating a new mint
        // (for PDA derivation of CMint account)
        // CompressAndCloseCMint does NOT need mint_signer - it verifies CMint via compressed_mint.metadata.mint
        if has_decompress_mint && !is_creating_mint {
            config = config.with_mint_signer(params.mint_seed);
        }
    }

    // Get account metas
    let account_metas = config.to_account_metas();

    // Serialize instruction data
    let data = instruction_data
        .data()
        .map_err(|e| RpcError::CustomError(format!("Failed to serialize instruction: {:?}", e)))?;

    // Build final instruction
    Ok(Instruction {
        program_id: LIGHT_TOKEN_PROGRAM_ID.into(),
        accounts: account_metas,
        data,
    })
}

/// Parameters for decompressing a mint to a CMint Solana account.
/// CMint is always compressible.
#[derive(Debug, Clone)]
pub struct DecompressMintParams {
    /// Rent payment in epochs (prepaid). Must be >= 2.
    pub rent_payment: u8,
    /// Lamports allocated for future write operations (top-up per write).
    pub write_top_up: u32,
}

impl Default for DecompressMintParams {
    fn default() -> Self {
        Self {
            rent_payment: 2, // Minimum valid rent_payment
            write_top_up: 0, // No write top-up by default
        }
    }
}

/// Helper function to create a comprehensive mint action instruction
#[allow(clippy::too_many_arguments)]
pub async fn create_comprehensive_mint_action_instruction<R: Rpc + Indexer>(
    rpc: &mut R,
    mint_seed: &Keypair,
    authority: Pubkey,
    payer: Pubkey,
    // Whether to decompress the mint to a CMint Solana account (with rent params)
    decompress_mint: Option<DecompressMintParams>,
    mint_to_recipients: Vec<(Pubkey, u64)>,
    update_mint_authority: Option<Pubkey>,
    update_freeze_authority: Option<Pubkey>,
    // Parameters for mint creation (required when creating a new mint)
    new_mint: Option<NewMint>,
) -> Result<Instruction, RpcError> {
    // Derive addresses
    let address_tree_pubkey = rpc.get_address_tree_v2().tree;
    let compressed_mint_address =
        derive_cmint_compressed_address(&mint_seed.pubkey(), &address_tree_pubkey);

    // Build actions
    let mut actions = Vec::new();

    if !mint_to_recipients.is_empty() {
        let recipients = mint_to_recipients
            .into_iter()
            .map(|(recipient, amount)| MintToRecipient { recipient, amount })
            .collect();

        actions.push(MintActionType::MintTo {
            recipients,
            token_account_version: 2, // V2 for batched merkle trees
        });
    }

    if let Some(new_authority) = update_mint_authority {
        actions.push(MintActionType::UpdateMintAuthority {
            new_authority: Some(new_authority),
        });
    }

    if let Some(new_authority) = update_freeze_authority {
        actions.push(MintActionType::UpdateFreezeAuthority {
            new_authority: Some(new_authority),
        });
    }

    // Add DecompressMint action if requested
    if let Some(decompress_params) = decompress_mint {
        let (_, cmint_bump) = find_cmint_address(&mint_seed.pubkey());
        actions.push(MintActionType::DecompressMint {
            cmint_bump,
            rent_payment: decompress_params.rent_payment,
            write_top_up: decompress_params.write_top_up,
        });
    }

    create_mint_action_instruction(
        rpc,
        MintActionParams {
            compressed_mint_address,
            mint_seed: mint_seed.pubkey(),
            authority,
            payer,
            actions,
            new_mint,
        },
    )
    .await
}

File: js/token-kit/src/codecs/mint-action.ts
/**
 * MintAction instruction codecs using Solana Kit patterns.
 *
 * Handles encoding of MintAction instruction data (discriminator 103).
 * Uses manual Borsh encoding via DataView/Uint8Array for complex nested
 * structures, following the same approach as transfer2.ts.
 */

import type { ReadonlyUint8Array } from '@solana/codecs';

import {
    writeU8,
    writeU16,
    writeU32,
    writeU64,
    writeBool,
    writeOption,
    writeVecBytes,
    concatBytes,
} from './borsh-helpers.js';

import type {
    CompressedProof,
    ExtensionInstructionData,
} from './types.js';

import { encodeExtensionInstructionData } from './transfer2.js';
import { DISCRIMINATOR } from '../constants.js';

// ============================================================================
// MINT ACTION TYPES
// ============================================================================

/** Recipient for MintToCompressed action. */
export interface MintRecipient {
    /** Recipient pubkey (32 bytes). */
    recipient: ReadonlyUint8Array;
    /** Amount to mint. */
    amount: bigint;
}

/** Mint compressed tokens to compressed accounts. */
export interface MintToCompressedAction {
    type: 'MintToCompressed';
    /** Token account version. */
    tokenAccountVersion: number;
    /** Recipients to mint to. */
    recipients: MintRecipient[];
}

/** Mint tokens from a compressed mint to a token Solana account. */
export interface MintToAction {
    type: 'MintTo';
    /** Index into remaining accounts for the recipient token account. */
    accountIndex: number;
    /** Amount to mint. */
    amount: bigint;
}

/** Update mint authority or freeze authority of a compressed mint. */
export interface UpdateAuthorityAction {
    type: 'UpdateMintAuthority' | 'UpdateFreezeAuthority';
    /** New authority (32 bytes), or null to revoke. */
    newAuthority: ReadonlyUint8Array | null;
}

/** Update a metadata field on a compressed mint. */
export interface UpdateMetadataFieldAction {
    type: 'UpdateMetadataField';
    /** Index of the TokenMetadata extension in the extensions array. */
    extensionIndex: number;
    /** Field type: 0=Name, 1=Symbol, 2=Uri, 3=Custom key. */
    fieldType: number;
    /** Empty for Name/Symbol/Uri, key string bytes for custom fields. */
    key: ReadonlyUint8Array;
    /** UTF-8 encoded value. */
    value: ReadonlyUint8Array;
}

/** Update metadata authority on a compressed mint. */
export interface UpdateMetadataAuthorityAction {
    type: 'UpdateMetadataAuthority';
    /** Index of the TokenMetadata extension in the extensions array. */
    extensionIndex: number;
    /** New authority (use zero bytes to set to None). */
    newAuthority: ReadonlyUint8Array;
}

/** Remove a metadata key from a compressed mint. */
export interface RemoveMetadataKeyAction {
    type: 'RemoveMetadataKey';
    /** Index of the TokenMetadata extension in the extensions array. */
    extensionIndex: number;
    /** UTF-8 encoded key to remove. */
    key: ReadonlyUint8Array;
    /** 0=false, 1=true - don't error if key doesn't exist. */
    idempotent: number;
}

/** Decompress a compressed mint to a Mint Solana account. */
export interface DecompressMintAction {
    type: 'DecompressMint';
    /** Rent payment in epochs (prepaid, must be >= 2). */
    rentPayment: number;
    /** Lamports allocated for future write operations. */
    writeTopUp: number;
}

/** Compress and close a Mint Solana account. */
export interface CompressAndCloseMintAction {
    type: 'CompressAndCloseMint';
    /** If non-zero, succeed silently when Mint doesn't exist or cannot be compressed. */
    idempotent: number;
}

/** Union of all MintAction variants. */
export type MintAction =
    | MintToCompressedAction
    | UpdateAuthorityAction
    | MintToAction
    | UpdateMetadataFieldAction
    | UpdateMetadataAuthorityAction
    | RemoveMetadataKeyAction
    | DecompressMintAction
    | CompressAndCloseMintAction;

// ============================================================================
// CREATE MINT & MINT METADATA
// ============================================================================

/** CreateMint data for new mint creation. */
export interface CreateMint {
    /** Placeholder for address trees (4 bytes, currently all zeros). */
    readOnlyAddressTrees: Uint8Array;
    /** Placeholder for root indices (4 x u16, currently all zeros). */
    readOnlyAddressTreeRootIndices: number[];
}

/** Light Protocol-specific metadata for compressed mints. */
export interface MintMetadata {
    /** Version for upgradability. */
    version: number;
    /** Whether the mint has been decompressed. */
    mintDecompressed: boolean;
    /** PDA derived from mintSigner (32 bytes). */
    mint: ReadonlyUint8Array;
    /** Signer pubkey used to derive the mint PDA (32 bytes). */
    mintSigner: ReadonlyUint8Array;
    /** Bump seed from mint PDA derivation. */
    bump: number;
}

/** Mint instruction data for creating or updating a mint. */
export interface MintInstructionData {
    /** Total supply of tokens. */
    supply: bigint;
    /** Number of base 10 digits to the right of the decimal place. */
    decimals: number;
    /** Light Protocol-specific metadata. */
    metadata: MintMetadata;
    /** Optional mint authority (32 bytes). */
    mintAuthority: ReadonlyUint8Array | null;
    /** Optional freeze authority (32 bytes). */
    freezeAuthority: ReadonlyUint8Array | null;
    /** Optional extensions for additional functionality. */
    extensions: ExtensionInstructionData[] | null;
}

// ============================================================================
// CPI CONTEXT
// ============================================================================

/** CPI context for mint action operations. */
export interface MintActionCpiContext {
    /** Whether to set the CPI context. */
    setContext: boolean;
    /** Whether this is the first set context call. */
    firstSetContext: boolean;
    /** Address tree index if create mint. */
    inTreeIndex: number;
    /** Input queue index. */
    inQueueIndex: number;
    /** Output queue index. */
    outQueueIndex: number;
    /** Token output queue index. */
    tokenOutQueueIndex: number;
    /** Index of the compressed account that should receive the new address. */
    assignedAccountIndex: number;
    /** Placeholder for multiple address trees (4 bytes). */
    readOnlyAddressTrees: Uint8Array;
    /** Address tree pubkey (32 bytes). */
    addressTreePubkey: ReadonlyUint8Array;
}

// ============================================================================
// TOP-LEVEL INSTRUCTION DATA
// ============================================================================

/** Full MintAction instruction data (discriminator 103). */
export interface MintActionInstructionData {
    /** Leaf index in the merkle tree (only set if mint already exists). */
    leafIndex: number;
    /** Whether to prove by index (only set if mint already exists). */
    proveByIndex: boolean;
    /** Root index for address or validity proof. */
    rootIndex: number;
    /** Maximum lamports for rent and top-up (u16::MAX = no limit, 0 = no top-ups). */
    maxTopUp: number;
    /** Only set when creating a new mint. */
    createMint: CreateMint | null;
    /** Actions to perform on the mint. */
    actions: MintAction[];
    /** Validity proof (optional). */
    proof: CompressedProof | null;
    /** CPI context (optional). */
    cpiContext: MintActionCpiContext | null;
    /** Mint data (optional, for create or full state). */
    mint: MintInstructionData | null;
}

// ============================================================================
// ACTION ENCODERS
// ============================================================================

/**
 * Borsh enum variant indices for the Action enum, matching the Rust definition.
 */
const ACTION_VARIANT = {
    MintToCompressed: 0,
    UpdateMintAuthority: 1,
    UpdateFreezeAuthority: 2,
    MintTo: 3,
    UpdateMetadataField: 4,
    UpdateMetadataAuthority: 5,
    RemoveMetadataKey: 6,
    DecompressMint: 7,
    CompressAndCloseMint: 8,
} as const;

function encodeMintRecipient(r: MintRecipient): Uint8Array {
    return concatBytes([
        new Uint8Array(r.recipient),
        writeU64(r.amount),
    ]);
}

function encodeMintToCompressedAction(
    action: MintToCompressedAction,
): Uint8Array {
    const parts: Uint8Array[] = [
        writeU8(action.tokenAccountVersion),
        writeU32(action.recipients.length),
    ];
    for (const r of action.recipients) {
        parts.push(encodeMintRecipient(r));
    }
    return concatBytes(parts);
}

function encodeUpdateAuthority(newAuthority: ReadonlyUint8Array | null): Uint8Array {
    return writeOption(newAuthority, (v: ReadonlyUint8Array) =>
        new Uint8Array(v),
    );
}

function encodeMintToAction(action: MintToAction): Uint8Array {
    return concatBytes([
        writeU8(action.accountIndex),
        writeU64(action.amount),
    ]);
}

function encodeUpdateMetadataFieldAction(
    action: UpdateMetadataFieldAction,
): Uint8Array {
    return concatBytes([
        writeU8(action.extensionIndex),
        writeU8(action.fieldType),
        writeVecBytes(action.key),
        writeVecBytes(action.value),
    ]);
}

function encodeUpdateMetadataAuthorityAction(
    action: UpdateMetadataAuthorityAction,
): Uint8Array {
    return concatBytes([
        writeU8(action.extensionIndex),
        new Uint8Array(action.newAuthority),
    ]);
}

function encodeRemoveMetadataKeyAction(
    action: RemoveMetadataKeyAction,
): Uint8Array {
    return concatBytes([
        writeU8(action.extensionIndex),
        writeVecBytes(action.key),
        writeU8(action.idempotent),
    ]);
}

function encodeDecompressMintAction(
    action: DecompressMintAction,
): Uint8Array {
    return concatBytes([
        writeU8(action.rentPayment),
        writeU32(action.writeTopUp),
    ]);
}

function encodeCompressAndCloseMintAction(
    action: CompressAndCloseMintAction,
): Uint8Array {
    return writeU8(action.idempotent);
}

function encodeAction(action: MintAction): Uint8Array {
    switch (action.type) {
        case 'MintToCompressed':
            return concatBytes([
                writeU8(ACTION_VARIANT.MintToCompressed),
                encodeMintToCompressedAction(action),
            ]);
        case 'UpdateMintAuthority':
            return concatBytes([
                writeU8(ACTION_VARIANT.UpdateMintAuthority),
                encodeUpdateAuthority(action.newAuthority),
            ]);
        case 'UpdateFreezeAuthority':
            return concatBytes([
                writeU8(ACTION_VARIANT.UpdateFreezeAuthority),
                encodeUpdateAuthority(action.newAuthority),
            ]);
        case 'MintTo':
            return concatBytes([
                writeU8(ACTION_VARIANT.MintTo),
                encodeMintToAction(action),
            ]);
        case 'UpdateMetadataField':
            return concatBytes([
                writeU8(ACTION_VARIANT.UpdateMetadataField),
                encodeUpdateMetadataFieldAction(action),
            ]);
        case 'UpdateMetadataAuthority':
            return concatBytes([
                writeU8(ACTION_VARIANT.UpdateMetadataAuthority),
                encodeUpdateMetadataAuthorityAction(action),
            ]);
        case 'RemoveMetadataKey':
            return concatBytes([
                writeU8(ACTION_VARIANT.RemoveMetadataKey),
                encodeRemoveMetadataKeyAction(action),
            ]);
        case 'DecompressMint':
            return concatBytes([
                writeU8(ACTION_VARIANT.DecompressMint),
                encodeDecompressMintAction(action),
            ]);
        case 'CompressAndCloseMint':
            return concatBytes([
                writeU8(ACTION_VARIANT.CompressAndCloseMint),
                encodeCompressAndCloseMintAction(action),
            ]);
    }
}

// ============================================================================
// STRUCT ENCODERS
// ============================================================================

function encodeCreateMint(data: CreateMint): Uint8Array {
    const parts: Uint8Array[] = [
        new Uint8Array(data.readOnlyAddressTrees),
    ];
    for (const idx of data.readOnlyAddressTreeRootIndices) {
        parts.push(writeU16(idx));
    }
    return concatBytes(parts);
}

function encodeMintMetadata(data: MintMetadata): Uint8Array {
    return concatBytes([
        writeU8(data.version),
        writeBool(data.mintDecompressed),
        new Uint8Array(data.mint),
        new Uint8Array(data.mintSigner),
        writeU8(data.bump),
    ]);
}

function encodeMintInstructionData(data: MintInstructionData): Uint8Array {
    const parts: Uint8Array[] = [
        writeU64(data.supply),
        writeU8(data.decimals),
        encodeMintMetadata(data.metadata),
        writeOption(data.mintAuthority, (v: ReadonlyUint8Array) =>
            new Uint8Array(v),
        ),
        writeOption(data.freezeAuthority, (v: ReadonlyUint8Array) =>
            new Uint8Array(v),
        ),
    ];

    // Option<Vec<ExtensionInstructionData>>
    if (data.extensions === null) {
        parts.push(new Uint8Array([0]));
    } else {
        parts.push(new Uint8Array([1]));
        parts.push(writeU32(data.extensions.length));
        for (const ext of data.extensions) {
            parts.push(encodeExtensionInstructionData(ext));
        }
    }

    return concatBytes(parts);
}

function encodeCompressedProof(proof: CompressedProof): Uint8Array {
    return concatBytes([
        new Uint8Array(proof.a),
        new Uint8Array(proof.b),
        new Uint8Array(proof.c),
    ]);
}

function encodeMintActionCpiContext(
    ctx: MintActionCpiContext,
): Uint8Array {
    return concatBytes([
        writeBool(ctx.setContext),
        writeBool(ctx.firstSetContext),
        writeU8(ctx.inTreeIndex),
        writeU8(ctx.inQueueIndex),
        writeU8(ctx.outQueueIndex),
        writeU8(ctx.tokenOutQueueIndex),
        writeU8(ctx.assignedAccountIndex),
        new Uint8Array(ctx.readOnlyAddressTrees),
        new Uint8Array(ctx.addressTreePubkey),
    ]);
}

// ============================================================================
// TOP-LEVEL ENCODER
// ============================================================================

/**
 * Encodes the full MintAction instruction data including discriminator (103).
 *
 * Borsh layout:
 * - discriminator: u8 (103)
 * - leaf_index: u32
 * - prove_by_index: bool
 * - root_index: u16
 * - max_top_up: u16
 * - create_mint: Option<CreateMint>
 * - actions: Vec<Action>
 * - proof: Option<CompressedProof>
 * - cpi_context: Option<CpiContext>
 * - mint: Option<MintInstructionData>
 */
export function encodeMintActionInstructionData(
    data: MintActionInstructionData,
): Uint8Array {
    const parts: Uint8Array[] = [
        // Discriminator
        writeU8(DISCRIMINATOR.MINT_ACTION),

        // Base fields
        writeU32(data.leafIndex),
        writeBool(data.proveByIndex),
        writeU16(data.rootIndex),
        writeU16(data.maxTopUp),

        // Option<CreateMint>
        writeOption(data.createMint, encodeCreateMint),

        // Vec<Action>
        writeU32(data.actions.length),
    ];

    for (const action of data.actions) {
        parts.push(encodeAction(action));
    }

    // Option<CompressedProof>
    parts.push(writeOption(data.proof, encodeCompressedProof));
    parts.push(writeOption(data.cpiContext, encodeMintActionCpiContext));
    parts.push(writeOption(data.mint, encodeMintInstructionData));

    return concatBytes(parts);
}
